core/src/HIP/Kokkos_HIP_ReduceScan.hpp

```diff
@@ -463,69 +449,88 @@ __device__ bool hip_single_inter_block_reduce_scan_impl(
 
     {
       void* const shared_ptr = shared_data + word_count.value * threadIdx.y;
-      /* reference_type shared_value = */ ValueInit::init(functor, shared_ptr);
+      /* reference_type shared_value = */ functor.init(
+          static_cast<pointer_type>(shared_ptr));
```

Comment:
Can't you declare `shared_ptr` with the right type?

Reply:
It would probably also be good to rename so as not to collide with the smart pointer type in the C++ standard library